Set state takes a function

leebyron opened the pull request:
This diff enables setState to accept a function in addition to a state partial. If you provide a function, it will be called with the up-to-date `state, props, context` as arguments.

This enables some nicer syntax for complex setState patterns:

If setState is doing an increment and wants to guarantee atomicy, you need a function:

```
this.setState(state => ({ number: state.number + 1 }));
```

This atomicy is particularly important if setState is called multiple times in a single frame of execution as the result of complex user actions. It's a tricky bug to chase down and difficult to determine how to fix when you find it. The current pattern of reaching into _pendingState relies on an implementation detail.

In this example: props.doAction() may result in your ancestor re-rendering and providing you with new props or context. If setState is called directly with an object literal referencing `this.props` or `this.context`, it will use the *old* version of those values, not the new value. Using a function solves for this case:

```
this.props.doAction();
this.setState((state, props) => ({ number: state.number * props.multiplier }));
```

leebyron commented:
Note: `ReactCompositeComponentState-test` will fail now because it relies on `_pendingState` to make assertions.

I'm not entirely certain how to fix because I'm not entirely certain what that test is actually testing.

brigand commented:
Also,

```js
var update = React.addons.update;

// ...
  updateFooBar(value){
      this.setState(state => update(state, {foo: {bar: {$set: 5}}});
  }
```

And swap that with your favorite immutable update method.

leebyron commented:
@brigand not quite in this case. I'm proposing to keep the same semantics of setState(). I can enable that effect in a separate pull built on this same idea using replaceState().

The semantics of setState() are to return a partial state patch to merge in.

leebyron commented:
@brigand - however:

```js
updateFooBar(value){
  this.setState(state => ({ fooBar: update(state.fooBar, {foo: {bar: {$set: 5}}}) });
}
```

sebmarkbage commented:
lgtm. Just needs to fix the existing test and add tests of the new functionality.

leebyron commented:
Suggestion on how to approach the broken test? What conditions do we need to test there?

—
Sent from Mailbox

On Fri, Jan 30, 2015 at 8:30 PM, Sebastian Markbåge
<notifications@github.com> wrote:

> lgtm. Just needs to fix the existing test and add tests of the new functionality.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/facebook/react/pull/2991#issuecomment-72297936

sebmarkbage commented:
I think that it is probably better to remove the piece from that test that introspects pending state. Then add another exhaustive test of this new setState-with-a-function. Then assert the sequencing of calling it in the various life cycles. As well as the sequencing of when the callbacks gets invoked. Including what this.state is at various life-cycles.

leebyron commented:
Just updated this pull to fix up ReactCompositeComponentState-test. It no longer inspects internal information and in order to test the same lifecycle of pending state, I added more setState() calls to ensure they're called in the correct order.

I think this should also cover the usages of providing a function.

sebmarkbage commented:
:thumbsup: Feel free to merge when you're comfortable.

syranide commented:
I'm curious, in which cases can props/context differ from the ones in `this`? Or are they just there for completeness (i.e. avoid accessing this).

vjeux commented:
Nice, glad my API suggestion was picked up :)

vsiao commented:
Fixes https://github.com/facebook/react/issues/122, yeah?

sebmarkbage commented:
@syranide If you receive props in the same batch as where you have a state update, then this gets executed before shouldComponentUpdate passing the new props and state, before the instance is update.

@leebyron Your unit test doesn't cover this scenario. ^ Might be worth adding.

leebyron commented:
Yes, let me add a unit test to illustrate that case before I merge

andreypopp commented:
Why not making this a separate method (for example `updateState`) rather than overloading `setState`?

leebyron commented:
I don't have a strong opinion. There are pros and cons to each. Smaller API surface area vs explicit type definition per method. Both are unambiguous. @sebmarkbage?

sebmarkbage commented:
We originally explored something like `enqueueTransaction` which would allow the `setState` within the transaction to keep the pseudo-OO looking API. However, when we realized that we couldn't update `this.state` and `this.props` at the point in the life-cycle where this operation must occur, we decided to go with something more monadic (which is what we really wanted all along anyway).

We evaluated a bunch of options such as `transact`, `enqueueStateUpdate` and `updateState`. I thought `updateState` was too confusing. There is nothing in those names to indicate the difference between the two signatures, `updateState` and `setState`. The only way to tell them apart is by some memorized knowledge. I would definitely screw that up and put the wrong argument with the wrong name.

@vjeux suggest that we just overload `setState` and that solved the problem.

We're also planning on overloading the API to accept a Promise anyway so this seems like a good start.

That said, I never really liked `setState` since it is deceptive. It is really scheduling or enqueuing the next state merge. I think that whatever name it has, the overloaded API makes sense.

leebyron commented:
I just added a new test which illustrates the values of props and state when dealing with the parent child relationship that led us to this, @sebmarkbage.
